dds/DCPS/NetworkAddress.h

```diff
@@ -0,0 +1,87 @@
+/*
+ *
+ *
+ * Distributed under the OpenDDS License.
+ * See: http://www.opendds.org/license.html
+ */
+
+#ifndef OPENDDS_DCPS_NETWORK_ADDRESS_H
+#define OPENDDS_DCPS_NETWORK_ADDRESS_H
+
+#include "dcps_export.h"
+
+#ifndef ACE_LACKS_PRAGMA_ONCE
+#  pragma once
+#endif /* ACE_LACKS_PRAGMA_ONCE */
+
+#include "Definitions.h"
+#include "PoolAllocator.h"
+
+#include "ace/INET_Addr.h"
+
+OPENDDS_BEGIN_VERSIONED_NAMESPACE_DECL
+
+namespace OpenDDS {
+namespace DCPS {
+
+class OpenDDS_Dcps_Export NetworkAddress {
+public:
+  NetworkAddress();
+  NetworkAddress(const NetworkAddress& addr);
+
+  explicit NetworkAddress(const char*);
+  NetworkAddress(uint16_t port, const char*);
+
+#if defined (ACE_HAS_WCHAR)
+  explicit NetworkAddress(const wchar_t*);
+  NetworkAddress(uint16_t port, const wchar_t*);
+#endif
+
+  explicit NetworkAddress(const ACE_INET_Addr& addr);
+
+  NetworkAddress& operator=(const NetworkAddress& rhs);
+  NetworkAddress& operator=(const ACE_INET_Addr& rhs);
+
+  bool operator==(const NetworkAddress& rhs) const;
+  bool operator!=(const NetworkAddress& rhs) const;
+
+  bool operator<(const NetworkAddress& rhs) const;
+
+  bool addr_bytes_equal(const NetworkAddress& rhs) const;
+
+  ACE_INET_Addr to_addr() const;
+  void to_addr(ACE_INET_Addr&) const;
+
+  int16_t get_type() const;
+  void set_type(int16_t type);
+
+  uint16_t get_port_number() const;
+  void set_port_number(uint16_t port);
+
+  bool is_any() const;
+  bool is_loopback() const;
+
+  bool is_private() const; // IPv4 only
+
+  bool is_uniquelocal() const; // IPv6 only
+  bool is_linklocal() const; // IPv6 only
+  bool is_sitelocal() const; // IPv6 only
+
+private:
+  union ip46
```

Comment:
Was it impossible to declare and ACE_INET_Addr and get a pointer to its internal data structure?  Then, `to_addr` can just return the internal ACE_INET_Addr but the memcmp function will still be more performant.

Reply:
I think the problem with this approach is that ACE_INET_Addr doesn't take pains to zero any padding for its internal data structure, so direct memcmp there is risky. So far, constructing ACE_INET_Addr based on this internal data structure seems pretty inexpensive, but if we're willing to eat the memory cost for storing an internal copy we could remove that construction cost as well.

Reply:
Right.  I'm assuming that it is possible to get a pointer to the guts of the ACE_INET_Addr.  If that's not possible, then my suggestion won't work and we are left with either constructing the ACE_INET_Addr on the fly or declaring one and eating the storage cost.  If it does, then we can still zero-pad and use memcmp.